src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java

```diff
@@ -1084,11 +1085,42 @@ private VarSymbol addParameter(String name, Type p, boolean genArg) {
     }
 
     private MethodType typeToMethodType(Type mt) {
+        return typeToMethodType(mt, null);
+    }
+
+    private MethodType typeToMethodType(Type mt, MethodType interfaceMethodType) {
```

Comment:
It is nice that we can put all the complexity in this method. I wonder if we could somehow make this more general - e.g. instead of detecting whether there is an intersection type - simply compare the erased signature of the function descriptor with the erased signature of the lambda method, and see if the two are compatible. I believe that, maybe, a resolution check might work (e.g. try to pass the erased parameter to the erased function descriptor and see if we get a resolution error or not). In `Lower` we do something similar - see `Lower::lookupMethod`.

Another option could be to always use the erased function descriptor type, and then add the required casts inside the lambda method implementation body (in most cases there would be no casts) - e.g. do not rely on the conversions handled by the lambda metafactory, but generate code that is more 1-1 with what the metafactory expects (and fix it up on the javac side). But doing this might change signature for some of the lambda bodies - which, if we take into account serializable lambdas, could be problematic.

Reply:
I think I prefer the second option, but I'm wondering if there are really cases when we need to insert any cast. It seem that at least at the place where we are making the changes all the needed casts in the lambda body has already being inserted. What do you think?